File: CPP/14. Longest Common Prefix.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;

class Solution {
public:
	string longestCommonPrefix(vector<string>& strs) {
		string prefix = "";
		if (strs.size() == 0) {
			return prefix;
		}

		auto minIt = min_element(strs.begin(), strs.end(),
								[](const string& a, const string& b) {
									return a.size() < b.size();
								});

		string shorttestStr = *minIt;
		for (int i = 0; i < shorttestStr.size(); i++) {
			char currentChar = shorttestStr[i];

			for (const string& str : strs) {
				if (str[i] != currentChar) {
					return prefix;
				}
			}

			prefix += currentChar;
		}
	}
};

int main() {
	Solution sol;
	vector<string> strs = { "flower","flow","fligth" };
	cout << sol.longestCommonPrefix(strs) << endl;
}

File: Python/14. Longest Common Prefix.py
from typing import List

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        prefix=""
        if len(strs)==0: return prefix

        for i in range(len(min(strs))):
            c=strs[0][i]
            if all(a[i]==c for a in strs):
                prefix+=c
            else:
                break
        return prefix

if __name__ == '__main__':
    strs = ["flower","flow","flight"]
    result = Solution().longestCommonPrefix(strs)
    print(result)